pkg/smokescreen/conntrack/conn_tracker.go

```diff
@@ -15,19 +19,111 @@ type Tracker struct {
+// impact on statistics.
+type ConnSuccessRateTracker struct {
+	ConnAttempts         *cache.Cache
+	ConnSuccessRateStats atomic.Value
+}
+
+// ConnSuccessRateStats represents a timestamped output of computations performed over
+// connection attempts.
+//
+// ConnSuccessRate represents the proportion of unique domains whose most recent connection attempt
+// was successful, compared to the total number of domains with connection attempts.
+type ConnSuccessRateStats struct {
+	CalculatedAt    time.Time
+	ConnSuccessRate float64
+	TotalConns      int
+}
+
+// StartNewConnSuccessRateTracker creates a new ConnSuccessRateTracker with a specific calculation interval at which
+// ConnSuccessRateStats will be recomputed, and a time window to calculate those statistics over.
+// - calculationInterval is how often statistics will be recomputed, in seconds.
+// - calculationWindow is the period that statistics will be calculated over, in seconds.
+// - cleanupInterval is how often expired items (e.g., items older than the calculationWindow) will be deleted from memory, in seconds.
+func StartNewConnSuccessRateTracker(calculationInterval time.Duration, calculationWindow time.Duration, cleanupInterval time.Duration) *ConnSuccessRateTracker {
+	newSuccessTracker := &ConnSuccessRateTracker{
+		ConnAttempts: cache.New(calculationWindow, time.Second*cleanupInterval),
+	}
+	newSuccessTracker.ConnSuccessRateStats.Store(ConnSuccessRateStats{CalculatedAt: time.Now(), ConnSuccessRate: 100, TotalConns: 0})
+
+	go func() {
+		for {
+			var total, succeeded int
+			for _, success := range newSuccessTracker.ConnAttempts.Items() {
+				total++
+				if success.Object.(bool) {
+					succeeded++
+				}
+			}
+			var successRate float64
+			// Avoid divide by zero errors
+			if total == 0 {
+				successRate = float64(100)
+			} else {
+				successRate = (float64(succeeded) / float64(total)) * 100
+			}
+			newSuccessTracker.ConnSuccessRateStats.Store(ConnSuccessRateStats{CalculatedAt: time.Now(), ConnSuccessRate: successRate, TotalConns: total})
+
+			time.Sleep(calculationInterval)
+		}
+	}()
+
+	return newSuccessTracker
+}
+
+func NewTracker(idle time.Duration, statsc statsd.ClientInterface, logger *logrus.Logger, sd atomic.Value, successRateTracker *ConnSuccessRateTracker) *Tracker {
 	return &Tracker{
-		Map:          &sync.Map{},
-		ShuttingDown: sd,
-		Wg:           &sync.WaitGroup{},
-		IdleTimeout:  idle,
-		statsc:       statsc,
+		Map:                &sync.Map{},
+		ShuttingDown:       sd,
+		Wg:                 &sync.WaitGroup{},
+		IdleTimeout:        idle,
+		statsc:             statsc,
+		SuccessRateTracker: successRateTracker,
+	}
+}
+
+// RecordAttempt stores the result of the most recent connection attempt for a destination.
+func (tr *Tracker) RecordAttempt(dest string, success bool) {
+	if tr.SuccessRateTracker == nil {
+		return
+	}
+	tr.SuccessRateTracker.ConnAttempts.Set(normalizeDomainName(dest), success, cache.DefaultExpiration)
```

Comment:
This feels like the right place to add hostname normalization, although I wonder if we'll eventually want to be able to toggle normalization (either globally or for a specific `RecordAttempt()`) in case the underlying library is getting it wrong. Anyway, not blocking for now!

Reply:
Ack!